### PR TITLE
Profit based on actual payout if the payout is known

### PR DESCRIPTION
**Pull request by da-kami**
Prior to adding the fees on top of the profit and loss calculation I thought it's probably a good idea to finally fix some tech debt. 

The profit and loss should always be based on the payout, and if we know the specific payout we should use that instead of calculating from a price.

### CODE REVIEW

File: taker-frontend/src/types.ts [hunk 1]
```diff
@@ -58,8 +58,10 @@ export interface Cfd {
 
     margin: number;
 
-    profit_btc?: number;
-    profit_percent?: number;
+    profit_btc: number;
+    profit_percent: number;
+
+    payout: number;
```
Comment (thomaseizinger) [inline taker-frontend/src/types.ts]: This commit adds the field here but it is not yet present in the backend? :)

File: daemon/src/projection.rs [hunk 2]
```diff
@@ -208,18 +220,76 @@ pub struct Cfd {
...
+            role,
+        }
+    }
+
+    fn payout(self) -> Option<Amount> {
+        if let Some((tx, script)) = self.collab_settlement_tx {
+            return Some(extract_payout_amount(tx, script));
+        }
+
+        match (self.cet, self.timelocked_cet) {
```
Comment (thomaseizinger) [inline daemon/src/projection.rs]: You can `?` all of these and it will bubble up the `None`.

File: daemon/src/projection.rs [hunk 3]
```diff
@@ -208,18 +220,76 @@ pub struct Cfd {
...
+            collab_settlement_tx: None,
+            cet: None,
+            timelocked_cet: None,
+            commit_published: false,
+            refund_published: false,
+            role,
+        }
+    }
+
+    fn payout(self) -> Option<Amount> {
```
Comment (thomaseizinger) [inline daemon/src/projection.rs]: This could take `role` as a parameter to avoid the field.

**Comment by da-kami** [comment]
I noticed that there are still some todos in the code - would be great if we could split up work on fixing those (if we want them to be fixed). I am rebasing my prep-work for the funding-fee changes on top of this now to be able to estimate how much will have to be adapted :)

File: daemon/src/model/cfd.rs [hunk 4]
```diff
@@ -1234,20 +1234,31 @@ pub fn calculate_long_liquidation_price(leverage: Leverage, price: Price) -> Pri
...
+    margin: SignedAmount,
+) -> Result<(SignedAmount, Percent)> {
+    let profit = payout - margin;
+    let percent = Decimal::from_f64(100. * profit.as_sat() as f64 / margin.as_sat() as f64)
+        .context("Unable to compute percent")?;
+
+    Ok((profit, Percent(percent)))
+}
+
+/// Returns the Profit/Loss (P/L) as Bitcoin. Losses are capped by the provided margin
```
Comment (bonomat) [inline daemon/src/model/cfd.rs]: This description does not seem to be up2date. Do you mind to add a proper doc?

File: daemon/src/projection.rs [hunk 5]
```diff
@@ -432,9 +497,28 @@ impl Cfd {
...
+                Ok((profit_btc, profit_percent)) => {
+                    (Some(profit_btc), Some(profit_percent.to_string()))
+                }
+                Err(err) => {
+                    tracing::error!(%payout, margin=%self.margin, "Profit calculation failed: {:#}", err);
+                    (None, None)
+                }
+            };
 
+            self.payout = Some(payout);
```
Comment (bonomat) [inline daemon/src/projection.rs:484]: I think we should be distinguishing between `realized` and `unrealized` profits and model it that way, e.g. call the fields `realized_profit` and `unrealized_profit`, `payout` and `predicted_payout` (or similar).

To decide whether to calculate `realized` or `unrealized` profits should depend on the state (a CFD is either `open` or `closed`). 
Right now the logic of how to calculate the profit (`realized` or `unrealized`) is spread into different places which I find hard to follow. 

We don't have to do this in this PR but I think the complexity here will bite us eventually.

File: daemon/src/projection.rs (hunk 2 above)
Comment (bonomat) [inline daemon/src/projection.rs]: Can you explain this logic please: 

From the code comments I understand that : 

- If `self.cet` is present, the `cet` should have been published.
- If `self.timelocked_cet` is present, the `cet` has not been published

What if we don't know which `cet` will be published, both are `None` right? 

If this is the case, then this function returns `None` which leads to `self.profit_btc = profit_btc;` not being set (line 501+).  
Am I correct and there is a problem here?

**Review by bonomat — CHANGES_REQUESTED** [review]
I'm not sure if there is a problem here. Please have a look.

File: daemon/src/projection.rs (hunk 2 above)
Comment (da-kami) [reply daemon/src/projection.rs]: The current logic is, that we set the profit and payout based on the current price when calling `projection::Cfd::new` and then *overwrite* that profit and payout in case we already know the payout when calling `projection::Cfd::apply` based on an event containing a transaction spending from `lock` (transactions are collected in `projection::Aggregated`). 
I think this design could be improved - it's not easy to understand what is happening at the moment. But refactoring this was not my concern in this PR.

To answer the question:

Returning `None` means that we don't have a fixed payout yet, but keep the initially set payout & p/l based on the current price. This is expected behavior according to the current logic flow.

File: daemon/src/projection.rs (hunk 5 above)
Comment (da-kami) [reply daemon/src/projection.rs:484]: I totally agree, but that would be a breaking API change. I tried to avoid that and keep what we have but make it more correct than in was before. We can go further and change the API as well, but then we will have to communicate that to users running the old version. 

If we accept breaking the API then we could even go further. We could distinguish more clearly between values that are fixed and values that are dynamic (e.g. based on current price). We could more clearly model "unrealized" (= open) and "realized" (= closed).

File: daemon/src/model/cfd.rs [hunk 6]
```diff
@@ -1234,20 +1234,31 @@ pub fn calculate_long_liquidation_price(leverage: Leverage, price: Price) -> Pri
...
 }
 
-/// Returns the Profit/Loss (P/L) as Bitcoin. Losses are capped by the provided margin
 pub fn calculate_profit(
+    payout: SignedAmount,
+    margin: SignedAmount,
+) -> Result<(SignedAmount, Percent)> {
+    let profit = payout - margin;
+    let percent = Decimal::from_f64(100. * profit.as_sat() as f64 / margin.as_sat() as f64)
+        .context("Unable to compute percent")?;
```
Comment (thomaseizinger) [inline daemon/src/model/cfd.rs]: We are we doing all these calculations in float and not in Decimal?

File: daemon/src/model/cfd.rs [hunk 7]
```diff
@@ -1234,20 +1234,31 @@ pub fn calculate_long_liquidation_price(leverage: Leverage, price: Price) -> Pri
...
+
+    Ok((profit, Percent(percent)))
+}
+
+/// Returns the Profit/Loss (P/L) as Bitcoin. Losses are capped by the provided margin
+pub fn calculate_profit_from_current_price(
```
Comment (thomaseizinger) [inline daemon/src/model/cfd.rs]: Nit:

```suggestion
pub fn calculate_profit_at_price(
```

File: daemon/src/model/cfd.rs [hunk 8]
```diff
@@ -1234,20 +1234,31 @@ pub fn calculate_long_liquidation_price(leverage: Leverage, price: Price) -> Pri
...
     initial_price: Price,
-    closing_price: Price,
+    current_price: Price,
```
Comment (thomaseizinger) [inline daemon/src/model/cfd.rs]: Nit: I wouldn't be upset if we find a better name here. The function doesn't necessarily know, which price it will be called with. Perhaps `target_price` is better? (From it being the calculation _target_?)

File: daemon/src/model/cfd.rs [hunk 9]
```diff
@@ -1234,20 +1234,31 @@ pub fn calculate_long_liquidation_price(leverage: Leverage, price: Price) -> Pri
...
+    let profit = payout - margin;
+    let percent = Decimal::from_f64(100. * profit.as_sat() as f64 / margin.as_sat() as f64)
+        .context("Unable to compute percent")?;
+
+    Ok((profit, Percent(percent)))
+}
+
+/// Returns the Profit/Loss (P/L) as Bitcoin. Losses are capped by the provided margin
+pub fn calculate_profit_from_current_price(
     initial_price: Price,
```
Comment (thomaseizinger) [inline daemon/src/model/cfd.rs]: I think we call this `opening_price` in many other places.

File: daemon/src/model/cfd.rs [hunk 10]
```diff
@@ -1234,20 +1234,31 @@ pub fn calculate_long_liquidation_price(leverage: Leverage, price: Price) -> Pri
...
     quantity: Usd,
     leverage: Leverage,
     position: Position,
-) -> Result<(SignedAmount, Percent)> {
+) -> Result<(SignedAmount, Percent, SignedAmount)> {
     let inv_initial_price =
         InversePrice::new(initial_price).context("cannot invert invalid price")?;
     let inv_closing_price =
-        InversePrice::new(closing_price).context("cannot invert invalid price")?;
+        InversePrice::new(current_price).context("cannot invert invalid price")?;
```
Comment (thomaseizinger) [inline daemon/src/model/cfd.rs]: Nit: Variable names not aligned.

File: daemon/src/projection.rs (hunk 5 above)
Comment (thomaseizinger) [reply daemon/src/projection.rs:484]: > If we accept breaking the API then we could even go further. We could distinguish more clearly between values that are fixed and values that are dynamic (e.g. based on current price). We could more clearly model "unrealized" (= open) and "realized" (= closed).

If you don't remove the old field but just add two new ones, it is not a breaking change :)

File: daemon/src/projection.rs [hunk 11]
```diff
@@ -432,9 +497,28 @@ impl Cfd {
 
         // If we don't have a dedicated closing price, keep the one that is set (which is
         // based on current price).
-        if let Some(closing_price) = self.aggregated.closing_price {
-            let (profit_btc, profit_percent) = self.compute_profit(closing_price);
+        // TODO: This clone can proably be optimized away
```
Comment (thomaseizinger) [inline daemon/src/projection.rs]: `Option::as_ref()` is likely your friend here.

**Review by thomaseizinger — COMMENTED** [review]
I think the commits should be squashed or more clearly separated.

File: daemon/src/projection.rs (hunk 5 above)
Comment (da-kami) [reply daemon/src/projection.rs:484]: After some more thought I am actually not sure the current UI has an advantage in a split API. We would have to introduce decision logic for "is realized already set otherwise use unrealized". 
I still think we should change the API, but I don't think it should be done as part of this PR. We could potentially also use two separate UI components for open and closed positions in the UI. I feel this needs a bit more thinking and should not be rushed as part of this work.

File: daemon/src/projection.rs (hunk 2 above)
Comment (da-kami) [reply daemon/src/projection.rs]: OK, but how would that help me here? The condition is "if either of them is set", if I questionmark one for them I will never be able to check the other. Am I missing something?

Comment (thomaseizinger) [reply daemon/src/projection.rs]: Ah I didn't see the `|`. Nevermind then, you can't use `?` then, correct!

File: daemon/src/model/cfd.rs (hunk 6 above)
Comment (da-kami) [reply daemon/src/model/cfd.rs]: Good question... I think this is some more legacy code that I just blindly copy pasted :/ 
I'll change it.

File: daemon/src/projection.rs (hunk 2 above)
Comment (thomaseizinger) [reply daemon/src/projection.rs]: What you can do is use `Option::or`: https://doc.rust-lang.org/std/option/enum.Option.html#method.or

```rust
let cet = self.cet.or(self.timelocked_cet)?;
```

File: daemon/src/projection.rs [hunk 12]
```diff
@@ -208,18 +220,76 @@ pub struct Cfd {
...
+        if let Some((tx, script)) = self.collab_settlement_tx {
+            return Some(extract_payout_amount(tx, script));
+        }
+
+        match (self.cet, self.timelocked_cet) {
+            (Some(tx), _) | (_, Some(tx)) => {
+                let dlc = self
+                    .latest_dlc
+                    .as_ref()
+                    .expect("dlc to be present when we have a cet");
```
Comment (thomaseizinger) [inline daemon/src/projection.rs]: This can use a `?` to bail if the DLC is not set.

File: daemon/src/projection.rs (hunk 3 above)
Comment (da-kami) [reply daemon/src/projection.rs:254]: Unfortunately it's not that simple, because in `apply` I don't know the role. We don't store the role in the `Cfd`. I could add it to the `Cfd` and avoid serialization, but then it's probably better to just keep it in the `Aggregate`.

File: daemon/src/model/cfd.rs (hunk 9 above)
Comment (da-kami) [reply daemon/src/model/cfd.rs]: this is not part of the changeset of this PR and it is called `initial_price` in `projection::Cfd` and `model::Cfd` so I will not change that here.

File: daemon/src/projection.rs (hunk 12 above)
Comment (da-kami) [reply daemon/src/projection.rs]: Only if I change the return type to be a `Result`. I think it is fair to assume that the `Dlc` is always set here.

File: daemon/src/model/cfd.rs (hunk 10 above)
Comment (da-kami) [reply daemon/src/model/cfd.rs]: Not sure what you mean by this comment.

File: daemon/src/model/cfd.rs [hunk 13]
```diff
@@ -1234,20 +1234,30 @@ pub fn calculate_long_liquidation_price(leverage: Leverage, price: Price) -> Pri
     price * leverage / (leverage + 1)
 }
 
+pub fn calculate_profit(payout: SignedAmount, margin: SignedAmount) -> (SignedAmount, Percent) {
+    let profit = payout - margin;
+
+    let profit_sats = Decimal::from(profit.as_sat());
+    let margin_sats = Decimal::from(margin.as_sat());
+    let percent = dec!(100) * profit_sats / margin_sats;
```
Comment (da-kami) [inline daemon/src/model/cfd.rs:1242]: @thomaseizinger is it OK to use the primitive impls over `checked_mul` and `checked_div` here?
Otherwise we would need to return a `Result` again.

Comment (thomaseizinger) [reply daemon/src/model/cfd.rs:1242]: They will panic on overflow / underflow in debug mode and overflow / underflow in release builds. If you are okay with that, then you can use them.

File: daemon/src/projection.rs [hunk 14]
```diff
@@ -211,12 +228,51 @@ struct Aggregated {
...
+    fn payout(self, role: Role) -> Option<Amount> {
+        if let Some((tx, script)) = self.collab_settlement_tx {
+            return Some(extract_payout_amount(tx, script));
+        }
+
+        let tx = self.cet.or(self.timelocked_cet)?;
+        let dlc = self
+            .latest_dlc
+            .as_ref()
+            .expect("dlc to be present when we have a cet");
```
Comment (thomaseizinger) [inline daemon/src/projection.rs:249]: You could `?` this instead of panicking.

File: daemon/src/projection.rs [hunk 15]
```diff
@@ -211,12 +228,51 @@ struct Aggregated {
...
+
+        Some(extract_payout_amount(tx, dlc.script_pubkey_for(role)))
+    }
+}
+
+/// Returns output if it can be found or zero amount
+///
+/// If we cannot find an output for our script we assume that we were liquidated.
+fn extract_payout_amount(tx: Transaction, script: Script) -> Amount {
+    let payout = match tx
```
Comment (thomaseizinger) [inline daemon/src/projection.rs]: This `match` can be replaced with a `map` and an `unwrap_or_else`.

File: daemon/src/projection.rs [hunk 16]
```diff
@@ -211,12 +228,51 @@ struct Aggregated {
...
+            return Some(extract_payout_amount(tx, script));
+        }
+
+        let tx = self.cet.or(self.timelocked_cet)?;
+        let dlc = self
+            .latest_dlc
+            .as_ref()
+            .expect("dlc to be present when we have a cet");
+
+        Some(extract_payout_amount(tx, dlc.script_pubkey_for(role)))
```
Comment (thomaseizinger) [inline daemon/src/projection.rs]: Would be more readable to extract some variables here.

**Review by thomaseizinger — APPROVED** [review]

File: daemon/src/model/cfd.rs (hunk 10 above)
Comment (thomaseizinger) [reply daemon/src/model/cfd.rs]: Line 1258 uses `inv_closing_price` and line 1259 uses `current_price`.

File: daemon/src/projection.rs (hunk 14 above)
Comment (da-kami) [reply daemon/src/projection.rs:249]: Questionmarking would give a wrong impression and swallow an error - it *is* an error if we don't have a DLC but get a CET. For now I think the `expect` is the clearest solution.

File: daemon/src/model/cfd.rs (hunk 13 above)
Comment (da-kami) [reply daemon/src/model/cfd.rs:1242]: I think these calculations with `SignedAmount` should always fit into the `Decimal` type. I don't think we can effectively run into an under-/overflow here. I'll keep as is.

File: daemon/src/projection.rs (hunk 14 above)
Comment (thomaseizinger) [reply daemon/src/projection.rs:249]: It is an option so the None will bubble up.

**Review by thomaseizinger — APPROVED** [review]
Thanks!

**Comment by da-kami** [comment]
I noticed that `closing_price` and `opening_price` are likely still better names for `calculate_profit_at_price`. Changing `closing` to `target` is a pain.

File: daemon/src/projection.rs (hunk 5 above)
Comment (bonomat) [reply daemon/src/projection.rs:484]: > I still think we should change the API, but I don't think it should be done as part of this PR.

this 👍

File: daemon/src/projection.rs (hunk 2 above)
Comment (bonomat) [reply daemon/src/projection.rs]: I missed that. Got it, thanks for the explanation.

**Review by bonomat — APPROVED** [review]
LGTM
